introduce Logging support

stephenplusplus opened the pull request:
Fixes #842
#### Must Reads
- Introduction to the Cloud Logging API: https://cloud.google.com/logging/docs/api/
- Docs: https://stephenplusplus.github.io/gcloud-node/#/docs/master/logging
#### To Dos
- [x] Make sure ACLs are properly set for Topics: https://github.com/GoogleCloudPlatform/gcloud-node/pull/865#discussion_r47936716
- [x] Enable Cloud Logging API on System Test Account
- [x] Document createSink.config.filter
- [x] Inherit from Service / ServiceObject
- **Support Sink destinations**
  - [x] Storage Bucket
  - [x] BigQuery Dataset
  - [x] Pub/Sub Topic
- **Docs**
  - [x] Readme
  - [x] JSDocs
  - [x] Site
- **Tests**
  - [x] System
  - Unit
    - [x] Index
    - [x] Log
    - [x] Sink
#### Preview

``` js
// Authenticating on a global-basis. You can also authenticate on a per-API-
// basis (see Authentication section above).
var gcloud = require('gcloud')({
  projectId: 'my-project',
  keyFilename: '/path/to/keyfile.json'
});

var logging = gcloud.logging();

// Create a sink using a Bucket as a destination.
var gcs = gcloud.storage();

logging.createSink('my-new-sink', {
  destination: gcs.bucket('my-sink')
}, function(err, sink) {});

// Write a critical entry to a log.
var syslog = logging.log('syslog');

var resource = {
  type: 'gce_instance',
  labels: {
    zone: 'global',
    instance_id: 3
  }
};

var entry = syslog.entry(resource, {
  delegate: process.env.user
});

syslog.critical(entry, function(err) {});

// Get all entries in your project.
logging.getEntries(function(err, entries) {
  if (!err) {
    // `entries` contains all of the entries from the logs in your project.
  }
});
```

stephenplusplus commented:
@VikramTiwari - your thoughts are most welcome if you have the chance to play around with this!

```sh
$ npm install --save stephenplusplus/gcloud-node#spp--logging
```

VikramTiwari commented:
Hey @stephenplusplus I am at Chrome Dev Summit for today and tomorrow, but I will be testing it by the end of week for sure. :+1:

stephenplusplus commented:
@VikramTiwari awesome! I'll look for you on the feed :stars:

stephenplusplus commented:
@jgeewax to prepare for this to merge, can you check that the Logging API is enabled for the Travis project?

stephenplusplus commented:
Also, any ideas for getting Travis to auth as a user? Many of our system tests require user-level, non-service account auth.

filipjs commented on lib/logging/index.js:
```diff
@@ -0,0 +1,543 @@
+  if (config.destination instanceof Dataset) {
+    var dataset = config.destination;
+
+    config.destination = format('{baseUrl}/projects/{pId}/datasets/{dId}', {
```
Datasets and topics (below) also need ACLs set up.
https://cloud.google.com/logging/docs/export/configure_export#setting_product_name_short_permissions_for_writing_exported_logs

stephenplusplus replied:
Thanks for catching that. I fixed this in https://github.com/GoogleCloudPlatform/gcloud-node/commit/77349e6f6b1daa467a2795ac49a4344cfe8d06b6

filipjs commented on lib/logging/index.js:
```diff
@@ -0,0 +1,543 @@
+  if (config.destination instanceof Topic) {
+    var topic = config.destination;
+
+    config.destination = format('{baseUrl}/{topicName}', {
```
Does topicName include project?
Full destination should be {baseUrl}/projects/{pId}/topics/{tId}'

stephenplusplus replied:
Yep, `topic.name` is its formatted name: https://github.com/GoogleCloudPlatform/gcloud-node/blob/1ac48c3177010f04c72acfcf7f36c76f45cadacc/lib/pubsub/topic.js#L216

filipjs commented on lib/logging/index.js:
```diff
@@ -0,0 +1,543 @@
+
+    var nextQuery = null;
+
+    if (resp.nextPageToken) {
```
v1beta3 doesn't support ListSinks pagination

stephenplusplus replied:
Are there plans to add pagination support?

filipjs replied:
Yes, it's added in V2

filipjs commented on lib/logging/log.js:
```diff
@@ -0,0 +1,415 @@
+var ServiceObject = require('../common/service-object.js');
+
+/**
+ * A log is a named collection of records or entries, each entry representing a
```
remove "records or" ?

filipjs commented on lib/logging/log.js:
```diff
@@ -0,0 +1,415 @@
+  ServiceObject.call(this, {
+    parent: logging,
+    baseUrl: '/logs',
+    id: this.name,
```
Shouldn't "unformattedName" be used here? 
That variable is not used otherwise

stephenplusplus replied:
This relates to how requests are patched upwards through our library, continuously appending pieces of the URL. I'm only passing `name` here, as it's already being appended to:

`...com/v1beta3/projects/{projectId}/logs/`

A user can refer to their log only by name ('my-log'), instead of a fully qualified name including the project ID, and it will work the same.

RE: `unformattedName` good catch, will remove.

filipjs replied:
id: this.name => name has to be already encoded, which might not be the case here if user forgets to do it.

stephenplusplus replied:
Fixed.

filipjs commented on lib/logging/log.js:
```diff
@@ -0,0 +1,415 @@
+ * }, function(err, apiResponse) {});
+ */
+Log.prototype.alert = function(entry, labels, callback) {
+  this.insert(Log.assignSeverityToEntries_(entry, 'ALERT'), labels, callback);
```
This doesn't match the example (there are no labels)?
Does "assignSeverityToEntries" work correctly with a single entry? (not sure how "arrify" works)

stephenplusplus replied:
> This doesn't match the example (there are no labels)?

I try to keep the examples as "look how simple this is" as possible, catering to the common use case. Is it more common to include labels with every entry?

> Does "assignSeverityToEntries" work correctly with a single entry? (not sure how "arrify" works)

Yes: https://github.com/sindresorhus/arrify/blob/master/index.js

filipjs replied:
What I meant with the first question is: shouldn't the function call in example take 3 arguments?
(because alert() takes 3 arguments). 
I guess this works because I was the same in tests

stephenplusplus replied:
Yes, the `labels` argument is marked as optional in the parameter descriptions, so I chose not to show it in the example for the reasons above.

filipjs commented on lib/logging/log.js:
```diff
@@ -0,0 +1,415 @@
+ * }, function(err, apiResponse) {});
+ */
+Log.prototype.critical = function(entry, labels, callback) {
+  var entries = Log.assignSeverityToEntries_(entry, 'CRITICAL');
```
Same comment as in ".alert" wrapper.
Also, this code can be changed into 1 line like there :)

filipjs commented on lib/logging/log.js:
```diff
@@ -0,0 +1,415 @@
+ *
+ * @resource [projects.logs.entries.write API Documentation]{@link https://cloud.google.com/logging/docs/api/ref/rest/v1beta3/projects.logs.entries/write}
+ *
+ * @param {object|object[]} entry - A log entry, or array of entries, to write.
```
Maybe it's worth adding that batching entries is highly recommended, as the is a QPS limit.

filipjs commented on lib/logging/log.js:
```diff
@@ -0,0 +1,415 @@
+ * @param {object} entry - An entry object.
+ */
+Log.prototype.formatEntry_ = function(entry) {
+  entry = extend(true, {}, entry);
```
What does this line do?

stephenplusplus replied:
Yep, so the user's given object isn't affected.

filipjs commented on lib/logging/sink.js:
```diff
@@ -0,0 +1,214 @@
+     *   }
+     * });
+     */
+    create: true,
```
no update method?

filipjs commented on lib/logging/sink.js:
```diff
@@ -0,0 +1,214 @@
+    }
+
+    self.request({
+      uri: '',
```
empty uri?

stephenplusplus replied:
Our library patches `request`s upwards through the inheritance model. `request.uri` in this context will resolve to `https://logging.googleapis.com/v1beta3/projects/{projectId}/sinks/{thisSinkName}`. Anything specified in `uri` would get appended to that base uri.

filipjs commented on system-test/logging.js:
```diff
@@ -0,0 +1,457 @@
+
+      it('should set metadata', function(done) {
+        var metadata = {
+          filter: 'name:*'
```
those are unit tests, or against a live system?
if the later, this should fail, as that is an invalid filter
(filter must match an existing LogEntry field).

stephenplusplus replied:
This is against the real backend. Interesting that it doesn't fail, I confirm that it does in fact update the `filter` on the response object: https://github.com/stephenplusplus/gcloud-node/blob/310017043fba58daad17834e0b95cff9ba615229/system-test/logging.js#L287

filipjs replied:
Oh yes, this change wasn't released yet. It should start to fail next week :)

stephenplusplus replied:
Is that part of a new API release?

filipjs replied:
No, this will fail in both API versions.

filipjs commented on system-test/logging.js:
```diff
@@ -0,0 +1,457 @@
+        .end(logEntry);
+    });
+  });
+
```
A test is missing that exercise the batching path

filipjs commented on test/logging/log.js:
```diff
@@ -0,0 +1,534 @@
+      log.insert(ENTRY, LABELS, assert.ifError);
+    });
+
+    it('should exec callback with only error and API response', function(done) {
```
Can you expand what is this testing?

stephenplusplus replied:
Sure, specifically, this line: https://github.com/stephenplusplus/gcloud-node/blob/310017043fba58daad17834e0b95cff9ba615229/lib/logging/log.js#L347 -- we want to make sure *only* two total arguments are passed to the user and they are the error and response body from the API request.

filipjs commented on test/logging/sink.js:
```diff
@@ -0,0 +1,200 @@
+      });
+    });
+
+    it('should localize the name', function() {
```
what is "localize" in this context?

stephenplusplus replied:
Store it in memory as a property on the created object. "sink.name"

filipjs commented on test/logging/sink.js:
```diff
@@ -0,0 +1,200 @@
+
+    it('should call set metadata', function(done) {
+      sink.setMetadata = function(metadata, callback) {
+        assert.strictEqual(metadata.filter, FILTER);
```
Add a check that destination is still there and unchanged

stephenplusplus replied:
This is a unit test, so it's strictly testing this body of code: https://github.com/stephenplusplus/gcloud-node/blob/310017043fba58daad17834e0b95cff9ba615229/lib/logging/sink.js#L158-L160

filipjs commented on test/logging/sink.js:
```diff
@@ -0,0 +1,200 @@
+  describe('setMetadata', function() {
+    var METADATA = { a: 'b', c: 'd' };
+
+    it('should refresh the metadata', function(done) {
```
what is "refresh"?

stephenplusplus replied:
Because the update (https://cloud.google.com/logging/docs/api/ref/rest/v1beta3/projects.sinks/update) is a PUT vs a PATCH, our API combines the current state of metadata with the input to `setMetadata`, combining them, effectively re-creating PATCH semantics.

Refresh means we get the latest state of the metadata before making the update. This test makes sure we do that.

filipjs commented:
Some initial comments.
One thing I would really suggest is to wait for new version API, to save the headache of updating everything twice (changes shouldn't be that big)

stephenplusplus commented:
Thanks for the review! I will be addressing the feedback and making the changes in the near future.

Regarding the new API version, is there a estimate on when that will be publicly available?

filipjs commented on test/logging/sink.js:
```diff
@@ -0,0 +1,200 @@
+    });
+
+    it('should make the correct request', function(done) {
+      var currentMetadata = { e: 'f', g: 'h' };
```
Thanks for the comments above!
In which case, maybe change currentMetadata to e.g. { 'c': 'e', 'g': 'h'} so the intersection of current and new metadata is not empty

stephenplusplus commented:
@filipjs thanks again for the review! I made many of the changes in a follow-up commit: https://github.com/stephenplusplus/gcloud-node/commit/40c860dde83f91ae13d338fee97c45612a92b546, please let me know if you catch anything else I can change.

I'm running into a problem creating sinks now. Hopefully it's a simple mistake on my part, but I'm seeing the "The caller does not have permission" error. Here is a request that's failing:

```
{ method: 'POST',
  uri: 'https://logging.googleapis.com/v1beta3/projects/nth-circlet-705/sinks',
  json:
   { destination: 'storage.googleapis.com/gcloud-logging-test7bb30900-99fb-11e5-ba9b-f7178516428f',
     name: 'gcloud-logging-test7cced940-99fb-11e5-ba9b-f7178516428f' },
  headers:
   { Authorization: 'Bearer {token}',
     'User-Agent': 'gcloud-node/0.25.1' } }
```

Any ideas?

filipjs commented:
Re: failing test.
Is the auth token refreshed between tests? Maybe it just expired?

stephenplusplus commented:
I don't believe so, even running the test in isolation fails. We also use [google-auth-library](https://github.com/google/google-auth-library-nodejs) which makes sure the token isn't expired before each request is sent.

filipjs commented:
V2 is out
https://cloud.google.com/logging/docs/api/introduction_v2

stephenplusplus commented:
@filipjs congrats on the release and thanks for letting me know. I'll get started on the migration.

stephenplusplus commented:
Migration complete. PTAL @filipjs. Here's the diff commit: 

I've created a new [`Entry`](https://stephenplusplus.github.io/gcloud-node/#/docs/master/logging/entry) type. Here's an example to write a new entry.

```js
var syslog = logging.log('syslog');

var resource = {
  type: 'gce_instance',
  labels: {
    zone: 'global',
    instance_id: 3
  }
};

var entry = syslog.entry(resource, {
  delegate: process.env.user
});

syslog.alert(entry, function(err, apiResponse) {
  if (!error) {
    // Log entry inserted successfully.
  }
});
```

Entry objects also come back from calling `logging.getEntries()` & `log.getEntries()`. To read the data, you can call `entry.data`.

Looking forward to your review. Thanks!

stephenplusplus commented:
@callmehiphop PTAL.

callmehiphop commented on README.md:
```diff
@@ -13,6 +13,7 @@ This client supports the following Google Cloud Platform services:
 * [Google BigQuery](#google-bigquery)
 * [Google Cloud Datastore](#google-cloud-datastore)
 * [Google Cloud DNS](#google-cloud-dns)
+* [Google Cloud Logging](#google-cloud-logging)
```
Should we denote that this is API is still in beta? I believe for pubsub we used to add the text *(beta)* after the link.

stephenplusplus replied:
Yeah, good call!

stephenplusplus replied:
Done!

callmehiphop commented on lib/logging/entry.js:
```diff
@@ -0,0 +1,140 @@
+ */
+function Entry(resource, data) {
+  if (!data) {
+    this.data = resource;
```
Kind of curious, is `resource.type` optional?

stephenplusplus replied:
It is not!

stephenplusplus replied:
To elaborate, when you write an entry to a log, you can use a common resource that is shared for all of the entries. That's why it's optional here, so that the user doesn't have to pass in the same resource object to every new entry

callmehiphop replied:
Ok, then correct me if I'm wrong, but if we're allowing for a single argument and setting it to `data`, wouldn't that imply that resource is optional?

callmehiphop replied:
Ah, ok.

callmehiphop commented on lib/logging/entry.js:
```diff
@@ -0,0 +1,140 @@
+ *
+ * var syslog = logging.log('syslog');
+ *
+ * var resource = {
```
I think an additional example of using an entry in its most simple form would be nice here. 

e.g.

```javascript
var entry = syslog.entry('thing1', 'thing2');
```

stephenplusplus replied:
Sadly, I'm not sure I can find one of those. This is pretty much how they're going to look.

callmehiphop replied:
I suppose I'm a little confused then, per https://github.com/GoogleCloudPlatform/gcloud-node/pull/865/files#diff-7d991e86dd4fed14a9532c6a8b99291cR242 it seemed like that was possible.

callmehiphop commented on lib/logging/log.js:
```diff
@@ -0,0 +1,531 @@
+ * Write a log entry with a severity of "ALERT".
+ *
+ * This is a simple wrapper around {module:logging/log#insert}. All arguments
+ * are the same as documented there.
```
Could we just do a `@mixin` here instead? It might be a little tedious to jump from method to method? Especially if the method has been deep linked into.

stephenplusplus replied:
That's a good idea, but our docs FE doesn't handle mixins on the method sharing method level. It was written for classes to extend classes. I'll change these methods to point to the correct method (these are all pointed to the old method "insert" which is now "write"), but maybe save the docs change for another issue or the common docs overhaul?

callmehiphop commented on lib/logging/log.js:
```diff
@@ -0,0 +1,531 @@
+
+/**
+ * Write log entries from a streaming object input.
+ *
```
Maybe a little bit of a nitpick and sort of unnecessary, but could we add a `@return` block to specify a stream is returned?

stephenplusplus replied:
Done.

callmehiphop commented on lib/logging/log.js:
```diff
@@ -0,0 +1,531 @@
+/**
+ * Write a log entry with a severity of "CRITICAL".
+ *
+ * This is a simple wrapper around {module:logging/log#insert}. All arguments
```
see https://github.com/GoogleCloudPlatform/gcloud-node/pull/865/files#r47702149

callmehiphop commented on lib/logging/log.js:
```diff
@@ -0,0 +1,531 @@
+ * Write a log entry with a severity of "DEBUG".
+ *
+ * This is a simple wrapper around {module:logging/log#insert}. All arguments
+ * are the same as documented there.
```
see https://github.com/GoogleCloudPlatform/gcloud-node/pull/865/files#r47702149

callmehiphop commented on lib/logging/log.js:
```diff
@@ -0,0 +1,531 @@
+/**
+ * Write a log entry with a severity of "ERROR".
+ *
+ * This is a simple wrapper around {module:logging/log#insert}. All arguments
```
see https://github.com/GoogleCloudPlatform/gcloud-node/pull/865/files#r47702149

callmehiphop commented on lib/logging/log.js:
```diff
@@ -0,0 +1,531 @@
+/**
+ * Write a log entry with a severity of "INFO".
+ *
+ * This is a simple wrapper around {module:logging/log#insert}. All arguments
```
see https://github.com/GoogleCloudPlatform/gcloud-node/pull/865/files#r47702149

callmehiphop commented on lib/logging/log.js:
```diff
@@ -0,0 +1,531 @@
+ *
+ * This is a simple wrapper around {module:logging/log#insert}. All arguments
+ * are the same as documented there.
+ *
```
see https://github.com/GoogleCloudPlatform/gcloud-node/pull/865/files#r47702149

callmehiphop commented on lib/logging/log.js:
```diff
@@ -0,0 +1,531 @@
+ * Write a log entry with a severity of "WARNING".
+ *
+ * This is a simple wrapper around {module:logging/log#insert}. All arguments
+ * are the same as documented there.
```
see https://github.com/GoogleCloudPlatform/gcloud-node/pull/865/files#r47702149

callmehiphop commented on lib/logging/log.js:
```diff
@@ -0,0 +1,531 @@
+ * @param {module:logging/entry|module:logging/entry[]} entry - A log entry, or
+ *     array of entries, to write.
+ * @param {object=} options - Configuration object.
+ * @param {object[]} labels - Labels to set on the log.
```
Are these supposed to be `options.labels` and `options.resource`?

stephenplusplus replied:
Yep!

callmehiphop commented:
@stephenplusplus LGTM! Did you want to squash this and/or wait for another review from @filipjs?

stephenplusplus commented:
Squashed! Let's wait for a little bit to see if we get a thumbs up from @filipjs.

jgeewax commented:
Is it worthwhile to create a separate issue to add a bit more color to the overview, defining what the different pieces are and what the expected usage should be like? (My eyes are still fresh so I'm avoiding reading too much about Sinks and Logs and etc, so that I can give useful feedback on whether our docs help new people go "oh, I get it")

stephenplusplus commented:
Definitely. It might be nice to get some common use cases from a Logging team member as well, so we can center some examples around them.

filipjs commented on README.md:
```diff
@@ -337,9 +338,61 @@ zone.createVM(name, { os: 'ubuntu' }, function(err, vm, operation) {
+// Create a sink using a Bucket as a destination.
+var gcs = gcloud.storage();
+
+dns.createSink('my-new-sink', {
```
What's "dns" here?

stephenplusplus replied:
A mistake! Thanks for catching that.

filipjs commented on README.md:
```diff
@@ -337,9 +338,61 @@ zone.createVM(name, { os: 'ubuntu' }, function(err, vm, operation) {
+  delegate: process.env.user
+});
+
+log.critical(entry, function(err) {});
```
log -> syslog?

filipjs commented on lib/logging/index.js:
```diff
@@ -0,0 +1,576 @@
+ *
+ * var config = {
+ *   destination: gcs.bucket('logging-bucket'),
+ *   filter: 'metadata.severity = ALERT'
```
In V2:
metadata.severity => severity

filipjs commented on lib/logging/index.js:
```diff
@@ -0,0 +1,576 @@
+ * //     }
+ * //   },
+ * //   jsonPayload: {
+ * //     delgate: process.env.USER
```
delgate => delegate

filipjs commented on lib/logging/index.js:
```diff
@@ -0,0 +1,576 @@
+ *
+ * @private
+ */
+Logging.prototype.setAclForTopic_ = function(name, config, callback) {
```
I need to check if this will work, as it differs from our docs.
I'll get back to you.

filipjs replied:
Yes, this will work.

filipjs commented on lib/logging/log.js:
```diff
@@ -0,0 +1,549 @@
+ * var log = logging.log('syslog');
+ */
+function Log(logging, name) {
+  var methods = {
```
There will be a "list logs" command added

stephenplusplus replied:
Where can I find the API docs? Looking here, I only see delete: https://cloud.google.com/logging/docs/api/ref_v2beta1/rest/v2beta1/projects.logs

filipjs replied:
There are no docs yet, but it's going to be pretty much similar to V1
https://cloud.google.com/logging/docs/api/ref/rest/v1beta3/projects.logs/list
with the service replaced with monitored resource

stephenplusplus replied:
I'm getting a 404 on https://logging.googleapis.com/v2beta1/projects/{projectId}/logs. Is this part of the next release for the API?

filipjs replied:
Yes. What I meant is that it's not ready yet, but it will be there soon

stephenplusplus replied:
I'll make an issue to support it in case this PR lands before then.

filipjs commented on lib/logging/log.js:
```diff
@@ -0,0 +1,549 @@
+    /**
+     * Delete the log.
+     *
+     * @resource [projects.logs.delete API Documentation]{@link https://cloud.google.com/logging/docs/api/ref/rest/v2beta1/projects.logs/delete}
```
Unfortunately the docs path is a little different from V1:
https://cloud.google.com/logging/docs/api/ref_v2beta1/rest/v2beta1/projects.logs/delete

filipjs commented on lib/logging/log.js:
```diff
@@ -0,0 +1,549 @@
+ * @return {string}
+ */
+Log.formatName_ = function(projectId, name) {
+  if (name.indexOf('logs/') > -1) {
```
This path returns a name which might not be escaped, e.g.:
projects/X/logs/a/b

stephenplusplus replied:
Does this look correct?:

```js
Log.formatName_ = function(projectId, name) {
  var path = 'projects/' + projectId + '/logs/';
  name = name.replace(path, '');

  if (decodeURIComponent(name) === name) {
    // The name has not been encoded yet.
    name = encodeURIComponent(name);
  }

  return path + '/' + name;
};
```

filipjs replied:
It would be safer to use a regexp? If name has different projectId, the replace won't match.

stephenplusplus replied:
All Log objects are scoped to a projectId that's provided up front to the parent Logging instance. Is this a possible scenario?

```js
var logging = gcloud.logging({ projectId: 'project-1' });
var log = logging.log('projects/project-2/logs/log-name');
```

filipjs replied:
Based on the method definition, yes. Nothing stops the user to give invalid data?
(unless the user cannot call this method directly, in which case why not say that name is only the  part after projects/*/logs/ ?

stephenplusplus replied:
The user cannot call this method directly, it's called when they access a Log object:

```js
var logging = gcloud.logging({ projectId: 'project' });
var log = logging.log('log-name'); // Logging.formatName_ is called behind the scenes here
```

> Based on the method definition, yes. 

I'm asking based on use case. Is there a user case to access a log from another project? Is it possible?

> why not say that name is only the part after projects/*/logs/

That's possible, but this caters to an API response which is in the full format. It also allows a user to use whichever version of the name (formatted or short) that they may be working with in other parts of their code.

filipjs commented on lib/logging/log.js:
```diff
@@ -0,0 +1,549 @@
+ * @param {object} callback.apiResponse - The full API response.
+ *
+ * @example
+ * var entry = log.entry('compute.googleapis.com', {
```
This code section still uses "services" (here in commends and in the method below).

filipjs commented on lib/logging/index.js:
```diff
@@ -0,0 +1,576 @@
+ * @param {object=} options - Services filtering options.
+ * @param {boolean} options.autoPaginate - Have pagination handled
+ *     automatically. Default: true.
+ * @param {number} options.pageSize - Maximum number of logs to return.
```
What are the default values if not provided in the options?

stephenplusplus replied:
They're not defined, so it lets the API decide.

filipjs commented on lib/logging/index.js:
```diff
@@ -0,0 +1,576 @@
+ *
+ * @resource [entries.list API Documentation]{@link https://cloud.google.com/logging/docs/api/ref_v2beta1/rest/v2beta1/entries/list}
+ *
+ * @param {object=} options - Services filtering options.
```
The most important option "filter" is missing from the options :)
Also it's rarely a good idea to pass an empty one, as it will return 30 days of logs starting from oldest ones.

stephenplusplus replied:
Good catch, and done! We'll default to `timestamp desc` now.

filipjs commented on lib/logging/entry.js:
```diff
@@ -0,0 +1,140 @@
+  var entry = extend(true, {}, this);
+
+  var whitelist = [
+    'logName',
```
last part of "logName" must be URL escaped, not sure if it's better to do it here, or in the constructor of Entry()

stephenplusplus replied:
For an Entry, it either comes from the API or a user. But when a user creates it or uses it to write to a log, we assign the log name automatically from the `formattedName_` property on the Log that the entry is passed through.

Does the API always return URL escaped names?

filipjs replied:
Yes it does

stephenplusplus replied:
Cool, I think we're covered then.

filipjs commented on lib/logging/index.js:
```diff
@@ -0,0 +1,584 @@
+    policy.bindings = arrify(policy.bindings);
+
+    policy.bindings.push({
+      role: 'roles/editor',
```
roles/pubsub.publisher is enough
https://cloud.google.com/pubsub/access_control?hl=en#tbl_roles

stephenplusplus replied:
Great! Updated.

callmehiphop commented:
Thanks @stephenplusplus! Also, thanks for the review @filipjs!
